### DISCUSSION TITLE
Suggestion: We should avoid merge commits.

### DISCUSSION THREAD

**Discussion by shailesh-vaidya** [Seagate/cortx#321] [General] [answered]
Just a suggestion :relaxed:
https://github.com/Seagate/cortx shows merge commits in history and its bit difficult to associate commit with task/ticket. We can merge PR using `Squash and merge` option with proper description.  So that we can identify changes done for particular ticket/task.  

Current git history graph looks like this,

```
[root@ssc-vm-0951 cortx]# git log --decorate --oneline --graph
*   21e5e49 (HEAD, main) Merge pull request #263 from Seagate/Saumya-Sunder-Contribution-Guide
|\
| *   9b7cb02 Merge branch 'main' into Saumya-Sunder-Contribution-Guide
| |\
| |/
|/|
* |   1aa39a0 Merge pull request #282 from Seagate/justinzw-patch-6
|\ \
| * \   6cb2816 Merge branch 'main' into justinzw-patch-6
| |\ \
| |/ /
|/| |
* | |   3cc38af Merge pull request #285 from Seagate/novium258-patch-1
|\ \ \
| * \ \   292fc6b Merge branch 'main' into novium258-patch-1
| |\ \ \
| |/ / /
|/| | |
* | | |   cee4f5b Merge pull request #301 from Seagate/Saumya-Sunder-patch-3
|\ \ \ \
| * \ \ \   d4fade9 Merge branch 'main' into Saumya-Sunder-patch-3
| |\ \ \ \
| |/ / / /
|/| | | |
* | | | |   92443ef Merge pull request #316 from Seagate/add-submodules
|\ \ \ \ \
| * | | | | f469fa9 refactor: add cortx-manager and cortx-management-portal as submodules
* | | | | |   3f09a00 Merge pull request #317 from Seagate/Updates-on-19
|\ \ \ \ \ \
| |/ / / / /
```

**Comment by huanghua78** [comment, marked as answer]
Hello Shailesh,

I also suggested to do "Squash and merge" for every PR.
Also the commit messages must contain what this particular commit does - a brief and meaningful message.

Best Regards,
Huang HUa
________________________________

**Reply by justinzw** [reply]
@shailesh-vaidya @huanghua78 

@johnbent and I moved this to our brand new github discussions area.

I think this is a good subject to "discuss" here.

**Reply by shailesh-vaidya** [reply]
We can make it strict by disabling `Allow merge commits` in repository settings :smiley:

![image](https://user-images.githubusercontent.com/63289460/94290049-4e6a4580-ff77-11ea-9b6b-16154c114e87.png)

**Reply by mukul-seagate11** [reply]
> We can make it strict by disabling `Allow merge commits` in repository settings 😃
> 
> ![image](https://user-images.githubusercontent.com/63289460/94290049-4e6a4580-ff77-11ea-9b6b-16154c114e87.png)

Thanks, as this is already initiated

**Comment by shailesh-vaidya** [comment]
@justinzw @mukul-seagate11  

Currently we have 76 branches in https://github.com/Seagate/cortx  :worried:  
```
[root@localhost cortx]# git branch -r | wc -l
76
[root@localhost  cortx]#
```
Seems we are not deleting branches after merge. Ideally we should move to `fork -> change -> PR` workflow so branches will not be created in parent repo.
For branch base workflow, branch should be deleted after merge.  We can also automate this by enabling `Automatically delete head branches` in https://github.com/Seagate/cortx-re/settings

![image](https://user-images.githubusercontent.com/63289460/95601107-a96c6400-0a70-11eb-8415-be271c6bad97.png)

**Reply by mukul-seagate11** [reply]
> @justinzw @mukul-seagate11
> 
> Currently we have 76 branches in https://github.com/Seagate/cortx 😟
> 
> ```
> [root@localhost cortx]# git branch -r | wc -l
> 76
> [root@localhost  cortx]#
> ```
> 
> Seems we are not deleting branches after merge. Ideally we should move to `fork -> change -> PR` workflow so branches will not be created in parent repo.
> For branch base workflow, branch should be deleted after merge. We can also automate this by enabling `Automatically delete head branches` in https://github.com/Seagate/cortx-re/settings
> 
> ![image](https://user-images.githubusercontent.com/63289460/95601107-a96c6400-0a70-11eb-8415-be271c6bad97.png)

yes sounds good & should be implemented but before automating this process we have to discuss/notify all components

**Reply by shailesh-vaidya** [reply]
@mukul-seagate11  Can we at least remove stale and merged branches from [cortx ](https://github.com/Seagate/cortx)repo - https://github.com/Seagate/cortx/branches/stale

**Reply by mukul-seagate11** [reply]
> @mukul-seagate11 Can we at least remove stale and merged branches from [cortx ](https://github.com/Seagate/cortx)repo - https://github.com/Seagate/cortx/branches/stale

yes we have to plan to remove the stale branches

**Comment by mukul-seagate11** [comment]
@shailesh-vaidya, We had enabled auto delete branches from cortx repo once they are merged as well as will initiate to delete the stale branches as well
